cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DefaultTaintTrackingImpl.qll

```diff
@@ -54,15 +55,31 @@ predicate predictableOnlyFlow(string name) {
     ]
 }
 
+class DefaultTaintTrackingOptions extends string {
+  DefaultTaintTrackingOptions() { this = "DefaultTaintTrackingOptions" }
+
+  predicate isUserInputDefault(Expr source, string cause) {
+    exists(FlowSource node | cause = node.getSourceType() |
+      node = DataFlow::exprNode(source)
+      or
+      node = DataFlow::definitionByReferenceNodeFromArgument(source)
+    )
+  }
```

Comment:
Have you looked at differences between this and results from `isUserDefault`?  I don't think they have to be identical, but having a lot in common (including `cause` strings `"read"`, `"fread"` etc) would give me confidence.

Reply:
Yes I've look at that: the format of `cause` changes slightly, but otherwise the results seem identical. DCA didn't flag up any result changes.

With respect to `cause` this will now use the `description` from the `hasRemoteFlowSource` (via `getSourceType`). This means that instead of a plain variable or function name we get a descriptive string. For example, in the case of `fread` we will `string read by fread` instead of just plain `fread`.

Side note: I think we might want to tweak the descriptions slightly (adding an indefinite article at the beginning of each of the descriptions).